contrib/natsio/src/test/java/com/google/cloud/dataflow/contrib/natsio/NatsIOTest.java

```diff
@@ -0,0 +1,145 @@
+		subject =  subjects.get(0);
+		
+		loop = (System.getProperty("loop") == null ? 30000 : Integer.parseInt(System.getProperty("loop")));
+		interval = (System.getProperty("interval") == null ? 0 : Integer.parseInt(System.getProperty("interval")));
+		consumers = (System.getProperty("consumers") == null ? 1 :Integer.parseInt(System.getProperty("consumers")));
+		producers = (System.getProperty("producers") == null ? 1 : Integer.parseInt(System.getProperty("producers")));
+		
+		props = new Properties();
+		props.setProperty("servers", servers);
+		if (queue != null)
+			props.setProperty("queue", queue);
+	}
+	
+	
+	private BlockingDataflowPipelineOptions buildOptions() {
+		BlockingDataflowPipelineOptions options = PipelineOptionsFactory.create().as(BlockingDataflowPipelineOptions.class);
+		options.setRunner(BlockingDataflowPipelineRunner.class);
+		options.setProject(project);
+		options.setStagingLocation(stagingLocation);
+		options.setNumWorkers(1);
+		
+		return options;
+	}
+	
+	@Test
+	public void simplePublish() {
+		BlockingDataflowPipelineOptions options = buildOptions();
+		options.setJobName("simple-publish-" + System.currentTimeMillis());
+		
+		Pipeline p = Pipeline.create(options);
+		p
+			.apply(Create.of(subject)).setCoder(StringUtf8Coder.of())
+			.apply(ParDo.of(new DoFn<String, KV<String,String>>() {
+				@Override
+				public void processElement(DoFn<String, KV<String, String>>.ProcessContext c) throws Exception {
+					c.output(KV.of(c.element(), "hello world!!!"));
+				}
+			}))
+			.apply(NatsIO.Write.to(props));
+		p.run();
+	}
+	
+	@Test
+	public void publishSubscribe() throws InterruptedException {
+		// Starting a consumer
+		DataflowPipelineOptions consumerOptions = PipelineOptionsFactory.create().as(DataflowPipelineOptions.class);
+		consumerOptions.setRunner(DataflowPipelineRunner.class);
+		consumerOptions.setProject(project);
+		consumerOptions.setStagingLocation(stagingLocation);
+		consumerOptions.setNumWorkers(1);
+		consumerOptions.setJobName("consumer-" + System.currentTimeMillis());
+		
+		Pipeline p1 = Pipeline.create(consumerOptions);
+		p1
+			.apply(NatsIO.Read.withMaxNumRecords(subject, maxRecords, props))
+			.apply(ParDo.of(new DoFn<KV<String,String>,String>() {
+				@Override
+				public void processElement(DoFn<KV<String, String>, String>.ProcessContext c) throws Exception {
+					c.output(c.element().getKey() + "," + c.element().getValue());
+				}
+			}));
+		p1.run();
+
+		// Starting a producer
+		BlockingDataflowPipelineOptions producerOptions = buildOptions();
+		producerOptions.setJobName("producer-" + System.currentTimeMillis());
+		
+		Pipeline p2 = Pipeline.create(producerOptions);
+		p2
+			.apply(Create.of(subject)).setCoder(StringUtf8Coder.of())
+			.apply(ParDo.of(new DoFn<String, String>() {
+				public void processElement(DoFn<String, String>.ProcessContext c) throws Exception {
```

Comment:
It would be good to actually make this a static class with a useful name, to explain what this `DoFn` does.

Then you can move `conn = Connection.connect(props)` to `startBundle` and can close `conn` in `finishBundle`, and that better matches Dataflow SDK style.